### PR TITLE
[Mobile] Cover block uploads failure UI

### PR DESCRIPTION
**Pull request by mkevins**
## Description
This PR adds UI for failed uploads in the Cover block, and also allows cancellation of ongoing uploads.

#### Related PRs

* `gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2150

## Testing steps

<details>
<summary>Upload failure UI</summary>

### Upload failure UI

1. Add a cover block to a post
2. Tap "Add Image or Video"
3. Tap "Choose from Device"
4. Turn on airplane mode :airplane: 
5. Pick an image to upload

**Expected:**

A failure indicator icon should appear in the upper-right-hand corner of the block:

Subsequently, tapping the background image should provide an option to retry the upload or remove the media:

</details>

<details>
<summary>Upload cancellation</summary>

### Upload cancellation

1. Add a cover block to a post
2. Tap "Add Image or Video"
3. Tap "Choose from Device"
4. Pick an image to upload (observe that the progress bar appears)
5. Before the upload finishes, tap the background image

**Expected:**

A dialog should be presented allowing you to cancel the upload:

</details>

### The `renderContent` prop

The states in `MediaUploadProgress` component can utilized by passing a render prop, _or_ via functional props bound to the component's internal state machine. Some of the block's UI (e.g. handling touch) is dependent on the component's internal state. We can wrap all of the dependent code within the `renderContent` prop to access this state. Alternatively, we can give the block component state, and updated it via the state change functional props (`onFinishMediaUploadWithSuccess`, etc.). I chose to use the latter implementation because it seemed more versatile. The `MediaUploadProgress` `renderContent` prop was made optional.

### Failure icon

I did not find the triangle icon used in the design mockup, so instead used the dashicon "warning" icon, which renders as a circle. The exclamation mark in that svg is actually not part of the shape, so to make it white requires setting the background to white in the icon container. I gave the container a border-radius (so it renders as a circle, instead of a square) and this results in a white circle around the icon. cc: @iamthomasbishop :wave: can you let me know if this looks ok?

<details>
<summary>Upload failed icon</summary>

cover-upload-failed-icon
</details>

<details>
<summary>Design mockup</summary>

cover-upload-failure-flow
</details>

Another notable deviation from the mockup is that the upload retry / remove dialog is not in the bottomsheet, but rather in a parent-app dialog (e.g. AlertDialog in Android). In theory, we _could_ implement a bottomsheet menu for these options, but I believe it might require some changes on the native side. Another thing to consider is that the other media blocks currently all use this same dialog. However, if we try to use the same dialog (the current state of this PR), we are missing the retry message (`'Failed to insert media.\nPlease tap for options.'`).

If not displaying the retry message is a blocker, one simple solution would be to add the retry message in a bottomsheet dialog that subsequently opens the already existing parent-app dialog. But this introduces "yet another tap" to the UX, so may not be desirable.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/cover/style.native.scss [hunk 1]
```diff
@@ -46,3 +46,28 @@
...
+	width: 24px;
+	height: 24px;
+	justify-content: center;
+	align-items: center;
+}
+
+.uploadFailedIcon {
+	// fill: #fff;
+	fill: #c34629;
+	color: #fff;
```
Comment (mkevins) [inline packages/block-library/src/cover/style.native.scss]: This is passed as `currentColor` in the dashicon svg.

Comment (mkevins) [reply packages/block-library/src/cover/style.native.scss]: This is not necessary, and has been removed.

**Comment by geriux** [comment]
Hey, @mkevins 👋 Started to review/test but I noticed that on iOS the upload progress bar is not showing (on Android it is) maybe there's a z-index issue?

**Comment by mkevins** [comment]
Hi @geriux :wave:  thanks for taking a look. I _did_ run into that on Android too, but solved it with this: https://github.com/WordPress/gutenberg/pull/21198/files#diff-dbbdd2c3421fbb78e7d6cc0b65c4e8b1R57. So, I wonder now whether something can be different on iOS that would make that view invisible :thinking: . Thanks for the feedback, I'll investigate it :+1:

**Comment by mkevins** [comment]
Hi @geriux, I added a `z-index` value for the upload progress to address that it was invisible on iOS. I gave a spot check to a few other flows too (image, gallery, etc.) to make sure they'll all still work as expected too. Ready for another test, when you can :smile: . Thanks!

**Review by geriux — APPROVED** [review]
LGTM! Tested both on iOS and Android, also double-checked the other blocks using the `MediaUploadProgress` were working as expected too.

Approving code and functionality but there's still pending design/UX approval.

Nice work @mkevins!

**Comment by iamthomasbishop** [comment]
@mkevins I've not been able to test this on a device (I can't see the additional media sources in the Android build [here](https://github.com/wordpress-mobile/WordPress-Android/pull/11541)), but looking at the screenshots above, I have some initial feedback.

> Subsequently, tapping the background image should provide an option to retry the upload or remove the media:

If there is only 1 failed item on the post, can we remove the `Remove All` button on this dialog? (might be a separate native thing?)

> Warning icon

This looks pretty good. Can you provide the following info:

* What is the size of icon itself
* What is the border-width of the white (or size of the container)? 
* What are the margins from the top-right?

**Comment by mkevins** [comment]
Hi @iamthomasbishop :wave: :smile:  thanks for the quick feedback!

>  looking at the screenshots above, I have some initial feedback.
> 
> > Subsequently, tapping the background image should provide an option to retry the upload or remove the media:
> 
> If there is only 1 failed item on the post, can we remove the `Remove All` button on this dialog? (might be a separate native thing?)
> 

This should be possible, but I believe will be needed on the native side. Since this will also be affecting other media blocks, maybe it's best that I open an issue for that one and address in a separate PR. Wdyt?

> > Warning icon
> 
> This looks pretty good. Can you provide the following info:
> 
> * What is the size of icon itself
> * What is the border-width of the white (or size of the container)?

The icon itself is 20px in svg, but expands to fill its container, which is 24px. The white "border" is actually coming from the svg image, and is 2px:

<details>
<summary>Warning icon svg image</summary>

image
</details>

However, that actually becomes _slightly_ larger than 2px since the total 20px svg expands to 24px (so roughly 2.4px).

> * What are the margins from the top-right?

The margins (actually padding in the outer container) are 8px from the top and right.

> * What red color are you using? (probably should be using `$red-50` (light mode) and `$red-30` (dark mode) from [Color Studio](https://color-studio.blog))

I just used the eye-drop tool to pick the color from your mockup (`#c34629`). I'll make this exact though :+1: .

**Comment by iamthomasbishop** [comment]
> Since this will also be affecting other media blocks, maybe it's best that I open an issue for that one and address in a separate PR. Wdyt?

Sounds good to me 👍 

> Warning icon

Ok, that sounds about right regarding sizing — the frame of the icon should be 24 and it's slightly inset on the grid 2pts on each side from the looks of it). Rather than giving a solid white outline, I'm wondering if we can use a semi-transparent white (about 18% alpha) with only 1px width. That would look like this:

<a href="https://user-images.githubusercontent.com/1200432/79399648-88a3d580-7f49-11ea-965c-2e747b4e10b3.png"></a>

For inspection, I've added this to the design docs [here](https://zeroheight.com/226850145/p/983139-error-indicator).

> The margins (actually padding in the outer container) are 8px from the top and right.

Perfect 👍 

> I just used the eye-drop tool to pick the color from your mockup (#c34629). I'll make this exact though 👍 .

Upon further inspection, can we use [red-40](https://color-studio.blog) on both light/dark modes? It sits nicely on both.

-------------

One more thing that I'm curious about. We are putting this indicator on the top-right of the cell, which feels fine, but I'm curious (assuming we end up using this same pattern for other errors on other blocks) if there could be a scenario where it conflicts with other actions or UI, such as the image-editing button that sits on the top-right of an image block? If that's the case, I think we can move to the bottom-left or bottom-right of the cell — top-left being reserved for media type badge (like GIFs).

**Comment by mkevins** [comment]
Hi @iamthomasbishop :wave: 

> > Warning icon
> 
> Ok, that sounds about right regarding sizing — the frame of the icon should be 24 and it's slightly inset on the grid 2pts on each side from the looks of it). Rather than giving a solid white outline, I'm wondering if we can use a semi-transparent white (about 18% alpha) with only 1px width. That would look like this:
> 

So far, the icons we've used only have 1 color (and a background color of the container). So in previous iterations shown above, the icon was getting it's white color from the background of the container, and the `fill` color was set to the color-studio reds. That's why we couldn't separate the inner "!" color from the "border" color (i.e. it wasn't really a border, but instead, the background was showing through, since the svg was inset by 2px).

To achieve your designs I've added some paths to the svg to allow independent coloring:

<details>
<summary>Error icon with semi-transparent border</summary>

cover-upload-failed-icon-semi-transparent-border
</details>

Ideally, icons will be centralized, but it's not immediately clear what the ideal pattern is for this case (multiple colors), since it seems most (maybe all?) of the "centralized" icons (`@wordpress/icons`, `Dashicon`, etc.) are used with only one fill color. So, this icon component is specific to native cover block for now.

> > I just used the eye-drop tool to pick the color from your mockup (#c34629). I'll make this exact though +1 .
> 
> Upon further inspection, can we use [red-40](https://color-studio.blog) on both light/dark modes? It sits nicely on both.
> 

:+1: Sounds good!

> One more thing that I'm curious about. We are putting this indicator on the top-right of the cell, which feels fine, but I'm curious (assuming we end up using this same pattern for other errors on other blocks) if there could be a scenario where it conflicts with other actions or UI, such as the image-editing button that sits on the top-right of an image block? If that's the case, I think we can move to the bottom-left or bottom-right of the cell — top-left being reserved for media type badge (like GIFs).

I'm not sure how likely that scenario is that we'll use the same pattern for other media blocks (all the others so far have their failure state in the center, since there's nothing obstructing the view there). I tried it with the icon in the lower right, and it felt strange to me that it was so close to those other icons (block settings and remove block). Here's what it looks like:

<details>
<summary>Error icon in lower right corner</summary>

cover-upload-failed-icon-lower-right
</details>

Maybe we can iterate on another PR? Wdyt?

**Comment by iamthomasbishop** [comment]
> To achieve your designs I've added some paths to the svg to allow independent coloring:

I think to simplify this, we can probably just remove the bordering — I hate to break up and customize an icon's construction just for something like this. Simply having the white container background that fills the empty space in the middle should be enough.

> I tried it with the icon in the lower right, and it felt strange to me

Totally agreed, I noticed this while playing around w/ the cell component in Figma. I think top-right is fine assuming we don't run into conflicts w/ other UI. In that case, we can always iterate by moving to the bottom-left, as you said. 👍

**Comment by mkevins** [comment]
Thanks @geriux and @iamthomasbishop for your feedback!

**Pull request merged**